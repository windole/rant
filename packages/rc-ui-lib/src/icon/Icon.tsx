import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
