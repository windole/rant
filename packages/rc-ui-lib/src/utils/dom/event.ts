import { MouseEvent, TouchEvent } from 'react';

export const stopPropagation = (event: TouchEvent | MouseEvent | Event): void =>
  event.stopPropagation();

export function preventDefault(
  event: TouchEvent | MouseEvent | Event,
  isStopPropagation?: boolean,
): void {
  /* istanbul ignore else */
  if (typeof event.cancelable !== 'boolean' || event.cancelable) {
    event.preventDefault();
  }

  if (isStopPropagation) {
    stopPropagation(event);
  }
}
