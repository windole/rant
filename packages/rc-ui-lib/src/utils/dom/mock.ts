export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect');
  spy.mockReturnValue(rect as DOMRect);
  return () => spy.mockRestore();
}

export function mockRect(vertical?: boolean): () => void {
  return mockGetBoundingClientRect({
    width: vertical ? 0 : 100,
    height: vertical ? 100 : 0,
    top: vertical ? 0 : 100,
    right: vertical ? 0 : 100,
    left: vertical ? 100 : 0,
    bottom: vertical ? 100 : 0,
  });
}
