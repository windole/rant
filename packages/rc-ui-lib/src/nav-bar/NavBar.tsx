import React, { useContext, CSSProperties, useRef } from 'react';
import classnames from 'classnames';
import Icon from '../icon';

import { getZIndexStyle } from '../utils';
import { BORDER_BOTTOM, HAPTICS_FEEDBACK } from '../utils/constant';
import useHeight from '../hooks/use-height';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import type { NavBarProps } from './PropsType';

const NavBar: React.FC<NavBarProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('nav-bar', prefixCls);

  const navBarRef = useRef<HTMLDivElement>();
  const height = useHeight(navBarRef);

  const onClickLeft = (event) => props.onClickLeft?.(event);
  const onClickRight = (event) => props.onClickRight?.(event);

  const renderLeft = () => {
    return [
      props.leftArrow && <Icon key="icon" className={classnames(bem('arrow'))} name="arrow-left" />,
      props.leftArea && (
        <span key="left-area" className={classnames(bem('text'))}>
          {props.leftArea}
        </span>
      ),
    ];
  };

  const renderRight = () => {
    return <span className={classnames(bem('text'))}>{props.rightArea}</span>;
  };

  const renderNavBar = () => {
    const { title, fixed, border, zIndex } = props;
    const style: CSSProperties = getZIndexStyle(zIndex);

    const hasLeft = props.leftArrow || props.leftArea;
    const hasRight = props.rightArea;
    return (
      <div
        ref={navBarRef}
        style={style}
        className={classnames(bem({ fixed, 'safe-area-inset-top': props.safeAreaInsetTop }), {
          [BORDER_BOTTOM]: border,
        })}
      >
        <div className={classnames(bem('content'))}>
          {hasLeft && (
            <div className={classnames(bem('left'), HAPTICS_FEEDBACK)} onClick={onClickLeft}>
              {renderLeft()}
            </div>
          )}
          <div className={classnames(bem('title'), 'van-ellipsis')}>{title}</div>
          {hasRight && (
            <div className={classnames(bem('right'), HAPTICS_FEEDBACK)} onClick={onClickRight}>
              {renderRight()}
            </div>
          )}
        </div>
      </div>
    );
  };

  const renderPlaceholder = () => (
    <div
      className={classnames(bem('placeholder'))}
      style={{ height: height ? `${height}px` : undefined }}
    >
      {renderNavBar()}
    </div>
  );

  if (props.fixed && props.placeholder) {
    return renderPlaceholder();
  }
  return renderNavBar();
};

NavBar.defaultProps = {
  title: '',
  leftArea: '',
  rightArea: '',
  leftArrow: false,
  border: true,
  fixed: false,
  placeholder: false,
  zIndex: 1,
  safeAreaInsetTop: false,
};

export default NavBar;
