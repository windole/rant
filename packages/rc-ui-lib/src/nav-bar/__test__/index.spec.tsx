import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import NavBar from '..';
import Icon from '../../icon';
import { mockGetBoundingClientRect } from '../../utils/dom/mock';

describe('NavBar', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
    jest.restoreAllMocks();
  });

  it('should render left area correct', async () => {
    wrapper = mount(<NavBar leftArea="speaker-s" title="foo" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render right area correct', async () => {
    wrapper = mount(<NavBar rightArea="speaker-s" title="foo" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render ReactNode icon correct', async () => {
    wrapper = mount(<NavBar leftArea={<Icon name="cart-o" color="#1989fa" />} title="foo" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render leftArrow correct', async () => {
    wrapper = mount(<NavBar leftArrow title="foo" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correct when set props fixed and placeholder', async () => {
    const restore = mockGetBoundingClientRect({ height: 50 });
    wrapper = mount(<NavBar fixed placeholder title="foo" />);
    expect(toJson(wrapper)).toMatchSnapshot();
    restore();
  });

  it('should emit click-left event when clicking left text', async () => {
    const onClickLeft = jest.fn();
    wrapper = mount(<NavBar title="foo" leftArea="left" onClickLeft={onClickLeft} />);
    const item = wrapper.find('.rc-nav-bar__left');
    await item.simulate('click');
    expect(onClickLeft).toHaveBeenCalledTimes(1);
  });

  it('should emit click-right event when clicking right text', async () => {
    const onClickRight = jest.fn();
    wrapper = mount(<NavBar title="foo" rightArea="right" onClickRight={onClickRight} />);
    const item = wrapper.find('.rc-nav-bar__right');
    await item.simulate('click');
    expect(onClickRight).toHaveBeenCalledTimes(1);
  });

  it('should have safe-area-inset-top class when using safe-area-inset-top prop ', async () => {
    wrapper = mount(<NavBar title="foo" safeAreaInsetTop />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should change z-index when using zIndex prop', async () => {
    wrapper = mount(<NavBar title="foo" zIndex={100} />);

    expect(wrapper.find('.rc-nav-bar').getDOMNode().style.zIndex).toEqual('100');
  });
});
