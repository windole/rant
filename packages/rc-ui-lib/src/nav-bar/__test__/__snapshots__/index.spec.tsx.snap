// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should have safe-area-inset-top class when using safe-area-inset-top prop  1`] = `
<NavBar
  border={true}
  fixed={false}
  leftArea=""
  leftArrow={false}
  placeholder={false}
  rightArea=""
  safeAreaInsetTop={true}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar rc-nav-bar--safe-area-inset-top rc-hairline--bottom"
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <div
      className="rc-nav-bar__content"
    >
      <div
        className="rc-nav-bar__title van-ellipsis"
      >
        foo
      </div>
    </div>
  </div>
</NavBar>
`;

exports[`NavBar should render ReactNode icon correct 1`] = `
<NavBar
  border={true}
  fixed={false}
  leftArea={
    <Icon
      color="#1989fa"
      name="cart-o"
      tag="i"
    />
  }
  leftArrow={false}
  placeholder={false}
  rightArea=""
  safeAreaInsetTop={false}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar rc-hairline--bottom"
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <div
      className="rc-nav-bar__content"
    >
      <div
        className="rc-nav-bar__left rc-haptics-feedback"
        onClick={[Function]}
      >
        <span
          className="rc-nav-bar__text"
          key="left-area"
        >
          <Icon
            color="#1989fa"
            name="cart-o"
            tag="i"
          >
            <Badge
              className="van-icon van-icon-cart-o"
              showZero={true}
              style={
                Object {
                  "color": "#1989fa",
                  "fontSize": undefined,
                }
              }
              tag="i"
            >
              <i
                className="rc-badge__wrapper van-icon van-icon-cart-o"
                style={
                  Object {
                    "color": "#1989fa",
                    "fontSize": undefined,
                  }
                }
              />
            </Badge>
          </Icon>
        </span>
      </div>
      <div
        className="rc-nav-bar__title van-ellipsis"
      >
        foo
      </div>
    </div>
  </div>
</NavBar>
`;

exports[`NavBar should render correct when set props fixed and placeholder 1`] = `
<NavBar
  border={true}
  fixed={true}
  leftArea=""
  leftArrow={false}
  placeholder={true}
  rightArea=""
  safeAreaInsetTop={false}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar__placeholder"
    style={
      Object {
        "height": "50px",
      }
    }
  >
    <div
      className="rc-nav-bar rc-nav-bar--fixed rc-hairline--bottom"
      style={
        Object {
          "zIndex": 1,
        }
      }
    >
      <div
        className="rc-nav-bar__content"
      >
        <div
          className="rc-nav-bar__title van-ellipsis"
        >
          foo
        </div>
      </div>
    </div>
  </div>
</NavBar>
`;

exports[`NavBar should render left area correct 1`] = `
<NavBar
  border={true}
  fixed={false}
  leftArea="speaker-s"
  leftArrow={false}
  placeholder={false}
  rightArea=""
  safeAreaInsetTop={false}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar rc-hairline--bottom"
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <div
      className="rc-nav-bar__content"
    >
      <div
        className="rc-nav-bar__left rc-haptics-feedback"
        onClick={[Function]}
      >
        <span
          className="rc-nav-bar__text"
          key="left-area"
        >
          speaker-s
        </span>
      </div>
      <div
        className="rc-nav-bar__title van-ellipsis"
      >
        foo
      </div>
    </div>
  </div>
</NavBar>
`;

exports[`NavBar should render leftArrow correct 1`] = `
<NavBar
  border={true}
  fixed={false}
  leftArea=""
  leftArrow={true}
  placeholder={false}
  rightArea=""
  safeAreaInsetTop={false}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar rc-hairline--bottom"
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <div
      className="rc-nav-bar__content"
    >
      <div
        className="rc-nav-bar__left rc-haptics-feedback"
        onClick={[Function]}
      >
        <Icon
          className="rc-nav-bar__arrow"
          key="icon"
          name="arrow-left"
          tag="i"
        >
          <Badge
            className="rc-nav-bar__arrow van-icon van-icon-arrow-left"
            showZero={true}
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
            tag="i"
          >
            <i
              className="rc-badge__wrapper rc-nav-bar__arrow van-icon van-icon-arrow-left"
              style={
                Object {
                  "color": undefined,
                  "fontSize": undefined,
                }
              }
            />
          </Badge>
        </Icon>
      </div>
      <div
        className="rc-nav-bar__title van-ellipsis"
      >
        foo
      </div>
    </div>
  </div>
</NavBar>
`;

exports[`NavBar should render right area correct 1`] = `
<NavBar
  border={true}
  fixed={false}
  leftArea=""
  leftArrow={false}
  placeholder={false}
  rightArea="speaker-s"
  safeAreaInsetTop={false}
  title="foo"
  zIndex={1}
>
  <div
    className="rc-nav-bar rc-hairline--bottom"
    style={
      Object {
        "zIndex": 1,
      }
    }
  >
    <div
      className="rc-nav-bar__content"
    >
      <div
        className="rc-nav-bar__title van-ellipsis"
      >
        foo
      </div>
      <div
        className="rc-nav-bar__right rc-haptics-feedback"
        onClick={[Function]}
      >
        <span
          className="rc-nav-bar__text"
        >
          speaker-s
        </span>
      </div>
    </div>
  </div>
</NavBar>
`;
